pkg/resources/kafka/headlessService.go

```diff
@@ -92,3 +97,95 @@ func (r *Reconciler) deleteHeadlessService() error {
 
 	return err
 }
+
+// deleteNonHeadlessServices deletes the all-broker service that was created for the current KafkaCluster
+// if there is any and also the service of each broker
+func (r *Reconciler) deleteNonHeadlessServices(ctx context.Context) error {
+	svc := corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Service",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.KafkaCluster.GetNamespace(),
+			Name:      fmt.Sprintf(kafkautils.AllBrokerServiceTemplate, r.KafkaCluster.GetName()),
+		},
+	}
+
+	err := r.Client.Delete(ctx, &svc)
+	if err != nil && client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// delete broker services
+	labelSelector := labels.NewSelector()
+	for k, v := range apiutil.LabelsForKafka(r.KafkaCluster.GetName()) {
+		req, err := labels.NewRequirement(k, selection.Equals, []string{v})
+		if err != nil {
+			return err
+		}
+		labelSelector = labelSelector.Add(*req)
+	}
+
+	// add "has label 'brokerId' to matching labels selector expression
+	req, err := labels.NewRequirement(v1beta1.BrokerIdLabelKey, selection.Exists, nil)
+	if err != nil {
+		return err
+	}
+	labelSelector = labelSelector.Add(*req)
+
+	var services corev1.ServiceList
+	err = r.Client.List(ctx, &services,
+		client.InNamespace(r.KafkaCluster.GetNamespace()),
+		client.MatchingLabelsSelector{Selector: labelSelector},
+	)
+
+	if err != nil {
+		return errors.WrapIfWithDetails(err, "failed to list services",
+			"namespace", r.KafkaCluster.GetNamespace(),
+			"label selector", labelSelector.String())
+	}
+
+	// if NodePort is used for any of the external listeners, the corresponding services need to remain
+	// so that clients from outside the Kubernetes cluster can reach the brokers
+	filteredSvcsToDelete := services
```

Comment:
I think we dont have to use another variable for this.
we can use the services all way long.
later: services = nonNodePortServices(services)
...

Reply:
yea...but having this little extra variable named `filteredSvcsToDelete` would make the implementation more readable

Reply:
It is clear enough to me services = nonNodePortServices(services).
It is just my opinion and a suggestion. Your solution is ok for me.

Reply:
We can also use here (https://github.com/banzaicloud/koperator/blob/71000a41df4fe3c1565217df9ff75ae9b9adbec6/pkg/resources/kafka/headlessService.go#L137)  name like: brokerServices
and  brokerServices = nonNodePortServices(brokerServices).

This https://github.com/banzaicloud/koperator/blob/71000a41df4fe3c1565217df9ff75ae9b9adbec6/pkg/resources/kafka/headlessService.go#L151

"services" already filtered by label selector so it is filteredSvcsToDelete

Reply:
> "services" already filtered by label selector so it is equal filteredSvcsToDelete

I guess I have a slightly different interpretation - we use the label selectors to query all of the services that match the labels (and they are stored in the `services` variable), then we filter them based on the external listener configuration, which results in `filteredSvcsToDelete` and these are the services that we will actually remove at the end